cdc/redo/reader/reader.go

```diff
@@ -0,0 +1,439 @@
+	var i uint64
+	for l.rowHeap.Len() != 0 && i < maxNumberOfEvents {
+		item := heap.Pop(&l.rowHeap).(*logWithIdx)
+		if item.data.Row.Row.CommitTs <= l.cfg.startTs {
+			continue
+		}
+		ret = append(ret, item.data.Row)
+		i++
+
+		rl := &model.RedoLog{}
+		err := l.rowReader[item.idx].Read(rl)
+		if err != nil {
+			if err != io.EOF {
+				return nil, err
+			}
+			continue
+		}
+
+		ld := &logWithIdx{
+			data: rl,
+			idx:  item.idx,
+		}
+		heap.Push(&l.rowHeap, ld)
+	}
+
+	return ret, nil
+}
+
+// ReadNextDDL ...
+func (l *LogReader) ReadNextDDL(ctx context.Context, maxNumberOfEvents uint64) ([]*model.RedoDDLEvent, error) {
+	select {
+	case <-ctx.Done():
+		return nil, errors.Trace(ctx.Err())
+	default:
+	}
+
+	l.ddlLock.Lock()
+	defer l.ddlLock.Unlock()
+
+	// init heap
+	if l.ddlHeap.Len() == 0 {
+		for i := 0; i < len(l.ddlReader); i++ {
+			rl := &model.RedoLog{}
+			err := l.ddlReader[i].Read(rl)
+			if err != nil {
+				if err != io.EOF {
+					return nil, err
+				}
+				continue
+			}
+
+			ld := &logWithIdx{
+				data: rl,
+				idx:  i,
+			}
+			l.ddlHeap = append(l.ddlHeap, ld)
+		}
+		heap.Init(&l.ddlHeap)
+	}
+
+	ret := []*model.RedoDDLEvent{}
+	var i uint64
+	for l.ddlHeap.Len() != 0 && i < maxNumberOfEvents {
+		item := heap.Pop(&l.ddlHeap).(*logWithIdx)
+		if item.data.DDL.DDL.CommitTs <= l.cfg.startTs {
+			continue
+		}
+
+		ret = append(ret, item.data.DDL)
+		i++
+
+		rl := &model.RedoLog{}
+		err := l.ddlReader[item.idx].Read(rl)
+		if err != nil {
+			if err != io.EOF {
+				return nil, err
+			}
+			continue
+		}
+
+		ld := &logWithIdx{
+			data: rl,
+			idx:  item.idx,
+		}
+		heap.Push(&l.ddlHeap, ld)
+	}
+
+	return ret, nil
+}
+
+// ReadMeta ...
+func (l *LogReader) ReadMeta(ctx context.Context) (checkpointTs, resolvedTs uint64, err error) {
+	select {
+	case <-ctx.Done():
+		return 0, 0, errors.Trace(ctx.Err())
+	default:
+	}
+
+	l.metaLock.Lock()
+	defer l.metaLock.Unlock()
+
+	if l.meta != nil {
+		return l.meta.CheckPointTs, l.meta.ResolvedTs, nil
+	}
+
+	files, err := ioutil.ReadDir(l.cfg.Dir)
+	if err != nil {
+		return 0, 0, cerror.WrapError(cerror.ErrRedoFileOp, errors.Annotate(err, "can't read log file directory"))
+	}
+
+	haveMeta := false
+	metaList := map[uint64]*common.LogMeta{}
+	var maxCheckPointTs uint64
+	for _, file := range files {
+		if filepath.Ext(file.Name()) == common.MetaEXT {
+			path := filepath.Join(l.cfg.Dir, file.Name())
+			fileData, err := os.ReadFile(path)
+			if err != nil {
+				return 0, 0, cerror.WrapError(cerror.ErrRedoFileOp, err)
+			}
+
+			l.meta = &common.LogMeta{}
+			_, err = l.meta.UnmarshalMsg(fileData)
+			if err != nil {
+				return 0, 0, cerror.WrapError(cerror.ErrRedoFileOp, err)
+			}
+
+			if !haveMeta {
+				haveMeta = true
+			}
+			metaList[l.meta.CheckPointTs] = l.meta
+			// since checkPointTs is guaranteed to increase always
+			if l.meta.CheckPointTs > maxCheckPointTs {
+				maxCheckPointTs = l.meta.CheckPointTs
+			}
```

Comment:
Does this mean there could exist multiple meta files?

Reply:
yes, if fail over happened between cdc clusters, could end up with multiple meta files